feat: first commit

armandleopold opened the pull request:
## Description of the changes

_In this section, explain **what and why** not how._

_In most cases, you can leave out details about how a change has been made. Code is generally self-explanatory in this regard (and if the code is so complex that it needs to be explained in prose, that’s what source comments are for). Just focus on making clear the reasons why you made the change in the first place, the way things worked before the change (and what was wrong with that), the way they work now, and why you decided to solve it the way you did._

armandleopold commented:
I plan to use metric-server soon.

lentidas commented:
Hello!

Thank you very much for your contribution, it is much appreciated :wink:

First things first, we are no longer including any modules inside this repository, even the cluster modules are now separated into their own repositories.
Secondly, we decided that we will not be supporting new cloud platforms for the time being, focusing mainly on the ones we currently use in company.

That said, your contribution is still much appreciated and we can list your module in our documentation page on something like _Community-maintained modules_.

What we propose is that you **create a separate repository in your personal/organization account**. You can take inspiration from our cluster modules ([EKS](https://github.com/camptocamp/devops-stack-module-cluster-eks), [SKS (in development)](https://github.com/camptocamp/devops-stack-module-cluster-sks/tree/ISDEVOPS-212-initial-implementation), [KinD](https://github.com/camptocamp/devops-stack-module-cluster-kind)). You can include your example also inside your cluster repository.

I'm happy to guide you in any way I can. I also have a few suggestions for your example:
- I see you used MinIO in order to provide S3 storage. This module is mainly for local development/testing deployments with KinD. OVH should have some bucket storage equivalent that you should use in production instead, **otherwise data loss is possible**.
- The previous point may change how you pass the storage credentials to the other modules, namely kube-prometheus-stack, Thanos and Loki. In each of those modules, you can see that they have different variants for each platform, and the variants call the main module on the root of the repository. **If you need to create your own variant, we propose you also include those inside your repository, each in a separate folder, and each variant can call our root modules.**
- Keycloak is also used for development purposes with KinD. Does OVH have an equivalent, like Cognito in AWS or Active Directory in Azure? You may still use Keycloak, but at the very least your should provide an external SQL database in production.

Let me know if you have any issues or if there are any points that may be clarified!

(I'll leave the PR open for the time being to continue the discussion.)

armandleopold commented:
Hello @lentidas,

Thanks for your answer.

That's ok for me, I will use your repo as an upstream to my company's version, and will much appreciate the continuous improvement you'd make along.

Thanks for the tips. I definitely will make good use of it and will make changes. Especially for the minio part and the keycloak. 

Do you have tips for integrating metric-server ? I couldn't make it work the first time. I'll retry, and make an issue if I find it impossible to make it work myself.

I also plan to make an "air-gaped" version of your stack deployable to client's proprietary infrastructure. I found it convenient that the KinD is the closest to that, or maybe you have some other ideas ? 

If you agree we can let the conversation open for a couple of weeks until i make a satisfying version of your stack on my company's repo ?

Thank you again for your great work.
Armand

lentidas commented:
Hello @armandleopold 

> Do you have tips for integrating metric-server ? I couldn't make it work the first time. I'll retry, and make an issue if I find it impossible to make it work myself.

We usually simply add the metrics server chart via our application module. You can check [this example](https://github.com/camptocamp/devops-stack/blob/75c60ca47417500b3b8403336f46dd9355852f8f/examples/kind/main.tf#L254). Note [on this line](https://github.com/camptocamp/devops-stack/blob/75c60ca47417500b3b8403336f46dd9355852f8f/examples/kind/main.tf#L267) that we pass the necessary argument to make the metrics-server on KinD. You should not need this configuration on a cloud provider if their Kubernetes clusters is deployed with proper certificates.

> I also plan to make an "air-gaped" version of your stack deployable to client's proprietary infrastructure. I found it convenient that the KinD is the closest to that, or maybe you have some other ideas ?

When talking about deploying an air-gaped version of our stack, personally I would not use KinD in production. Using KinD would mean you have your cluster entirely on the same machine/VM running on the container storage that is installed. There would be no resiliency, etc, etc. 

The way I see it, there is nothing stopping you from deploying a cluster either way you see fit. For example, you could have VMs running your nodes in an hypervisor ([example using Proxmox](https://theitbros.com/set-up-kubernetes-on-proxmox/)) along with a TrueNAS machine to provide the [persistent storage](https://jonathangazeley.com/2021/01/05/using-truenas-to-provide-persistent-storage-for-kubernetes/) and [S3 storage](https://www.truenas.com/docs/core/coretutorials/services/configurings3/). Heck, you could even use TrueNAS Scale to accomplish all this. **This is entirely theoretical on my part, I've not tested it and surely there would be another easier way to deploy Kubernetes clusters.**

The point is, as long as you have those components above, you can then pass the necessary parameters to the providers on the Terraform code and deploy the DevOps Stack modules. In this case, you should probably use the KinD variants of the modules.

I hope I've given you some ideas on which to work on :)

I'll leave the PR open for a few weeks, no worries.

armandleopold commented:
Very instructive, thanks a lot ;-)
Will look it out.

armandleopold commented:
Hey @lentidas , i updated my exemple :

* switched to OVH s3 storage instead of Minio
* add keycloak persistence
* seperate files for locals, s3, variables, 
* update terraform providers 
* update module dependency versions

lentidas commented:
Hey @armandleopold,

Sorry for getting back to you so late on this PR. Thank you very much for your contributions!

We discussed this today as a team and we still agree that your module like I said on one of the last comments:

> What we propose is that you create a separate repository in your personal/organization account. You can take inspiration from our cluster modules ([EKS](https://github.com/camptocamp/devops-stack-module-cluster-eks), [SKS (in development)](https://github.com/camptocamp/devops-stack-module-cluster-sks/tree/ISDEVOPS-212-initial-implementation), [KinD](https://github.com/camptocamp/devops-stack-module-cluster-kind)). You can include your example also inside your cluster repository.

If you do, you can get back to us and we will gladly reference it as an external module on our documentation, but we will still leave you in charge of maintaining the module.

armandleopold commented:
Hello @lentidas ok understood thanks
